core/src/main/scala/io/github/ghostbuster91/sttp/client3/json/circe/CirceProductCodecGenerator.scala

```diff
@@ -1,54 +1,159 @@
                 Encoder.$productEncoder(..${propNames.map(p =>
             Lit.String(p.v)
           )})(p => (..$extractors))"""
       }
     }
 
-  private def decoder(product: Product.Regular) =
+  private def decoder(product: Product.Regular, coproducts: List[Coproduct]) =
     CirceTypeProvider.DecoderTpe.map { decoderTpe =>
       val resultClassType = product.name.typeName
       val decoderName = product.name.toParam.asPrefix("Decoder")
-      val productDecoder = Term.Name(s"forProduct${product.properties.size}")
+      val discriminatorPropName =
+        determineDiscriminator(
+          product: Product.Regular,
+          coproducts: List[Coproduct]
+        ).map(t => t._1)
+
       val productElements =
-        product.properties.map(p => Lit.String(p.paramName.v))
-      q"""implicit lazy val $decoderName: $decoderTpe[$resultClassType] =
+        product.properties
+          .filterNot(p => discriminatorPropName.contains(p.paramName))
+          .map(p => Lit.String(p.paramName.v))
+
+      productElements.size match {
+        case 0 =>
+          val constDecoder = Term.Name(s"const")
+          q"""implicit lazy val $decoderName: $decoderTpe[$resultClassType] =
+                Decoder.$constDecoder(${product.name.term}.apply)"""
+        case size =>
+          val productDecoder = Term.Name(s"forProduct$size")
+          q"""implicit lazy val $decoderName: $decoderTpe[$resultClassType] =
                 Decoder.$productDecoder(..$productElements)(${product.name.term}.apply)"""
+      }
+    }
+
+  private def determineDiscriminator(
+      product: Product.Regular,
+      coproducts: List[Coproduct]
+  ): Option[(ParameterName, Lit)] =
+    product.parents match {
+      case parents if parents.nonEmpty =>
+        val parentCoproducts = coproducts.filter { coproduct: Coproduct =>
+          parents.contains(coproduct.name)
+        }
+        val discriminator =
+          parentCoproducts.flatMap(cp => cp.discriminator).headOption
+        discriminator match {
+          case Some(Discriminator.StringDsc(fieldName, mapping)) =>
+            (
+              product.properties
+                .find(p => p.paramName.term.value == fieldName)
+                .map(_.paramName),
+              mapping
+                .find { case (_, className) => className == product.name }
+                .map(_._1)
+                .map(Lit.String(_))
+            ).sequence
+          case Some(Discriminator.EnumDsc(fieldName, _, mapping)) =>
+            (
+              product.properties
+                .find(p => p.paramName.term.value == fieldName)
+                .map(_.paramName),
+              mapping
+                .find { case (_, className) => className == product.name }
+                .map(_._1)
+                .map {
+                  case EnumValue.StringEv(v) => Lit.String(v)
+                  case EnumValue.IntEv(v)    => Lit.Int(v)
+                }
```

Comment:
It might be nicer not to use the Literal here, but the EnumValue itself, however, this adds complexity and we chose not to.

Reply:
I think that is a good compromise. We are the ones who generate the enum so we should be fine. Do you mind adding a comment here about that decision?

Reply:
Added a comment.